server/store/layer_generators/main.go

```diff
@@ -0,0 +1,232 @@
+	code, err := generateLayer("TimerLayer", "timer_layer.go.tmpl")
+	if err != nil {
+		return err
+	}
+	formatedCode, err := format.Source(code)
+	if err != nil {
+		return err
+	}
+
+	err = os.MkdirAll("timerlayer", 0700)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(path.Join("timerlayer", "timerlayer.go"), formatedCode, 0600)
+}
+
+type methodParam struct {
+	Name string
+	Type string
+}
+
+type methodData struct {
+	Params  []methodParam
+	Results []string
+}
+
+type storeMetadata struct {
+	Name    string
+	Methods map[string]methodData
+}
+
+func extractMethodMetadata(method *ast.Field, src []byte) methodData {
+	params := []methodParam{}
+	results := []string{}
+	ast.Inspect(method.Type, func(expr ast.Node) bool {
+		if e, ok := expr.(*ast.FuncType); ok {
+			if e.Params != nil {
+				for _, param := range e.Params.List {
+					for _, paramName := range param.Names {
+						params = append(params, methodParam{Name: paramName.Name, Type: string(src[param.Type.Pos()-1 : param.Type.End()-1])})
+					}
+				}
+			}
+			if e.Results != nil {
+				for _, result := range e.Results.List {
+					results = append(results, string(src[result.Type.Pos()-1:result.Type.End()-1]))
+				}
+			}
+		}
+		return true
+	})
+	return methodData{Params: params, Results: results}
+}
+
+func extractStoreMetadata() (*storeMetadata, error) {
+	// Create the AST by parsing src.
+	fset := token.NewFileSet() // positions are relative to fset
+
+	file, err := os.Open("store.go")
+	if err != nil {
+		return nil, fmt.Errorf("unable to open store.go file: %w", err)
+	}
+	src, err := io.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	file.Close()
+	f, err := parser.ParseFile(fset, "", src, parser.AllErrors|parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	topLevelFunctionsToSkip := map[string]bool{
+		"BeginTx":    true,
+		"RollbackTx": true,
+		"CommitTx":   true,
+	}
+
+	metadata := storeMetadata{Methods: map[string]methodData{}}
+
+	ast.Inspect(f, func(n ast.Node) bool {
+		if x, ok := n.(*ast.TypeSpec); ok {
+			if x.Name.Name == "Store" {
+				for _, method := range x.Type.(*ast.InterfaceType).Methods.List {
+					methodName := method.Names[0].Name
+					if skip := topLevelFunctionsToSkip[methodName]; !skip {
+						metadata.Methods[methodName] = extractMethodMetadata(method, src)
+					}
+				}
+			}
+		}
+		return true
+	})
+
+	return &metadata, nil
+}
+
+func generateLayer(name, templateFile string) ([]byte, error) {
+	out := bytes.NewBufferString("")
+	metadata, err := extractStoreMetadata()
+	if err != nil {
+		return nil, err
+	}
+	metadata.Name = name
+
+	myFuncs := template.FuncMap{
+		"joinResults": func(results []string) string {
+			return strings.Join(results, ", ")
+		},
+		"joinResultsForSignature": func(results []string) string {
+			if len(results) == 0 {
+				return ""
+			}
+			returns := []string{}
+			for _, result := range results {
+				switch result {
+				case "*PostReminderMetadata":
```

Comment:
When does this case occur?

Reply:
It doesn't! Good catch -- copy and paste error on my part :)